### PR TITLE
Implemented Google Sign up and MVVM

### PR DESCRIPTION
**Pull request by mustiikhalil**
What is the PR about?

Implemented the Google on click sign up and log in. Started Using the MVVM Architecture for the project instead of MVC

Please update the pods, and get the credentials file for firebase :D

### CODE REVIEW

File: todoie/Bindable.swift [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+//  Bindable.swift
+//  todoie
+//
+//  Created by Mustafa Khalil on 1/2/19.
+//  Copyright © 2019 Aurora. All rights reserved.
+//
+
+import Foundation
+
+class Bindable<T>{
```
Comment (suhabaobaid) [inline todoie/Bindable.swift]: what is this used for exactly? How does it bind the viewcontroller with viewmodel?

Comment (mustiikhalil) [reply todoie/Bindable.swift]: @suhabaobaid read the following comments

File: todoie/Bindable.swift [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    
+    // oberver is the function that will be called when the value changes with the value
```
Comment (mustiikhalil) [inline todoie/Bindable.swift]: This is going to be the observer that should invoke the code that we set in the view controller.

File: todoie/Bindable.swift [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    fileprivate var observer: ((T?)->())?
+    
+    // value is the new value inputed to the bindable object
```
Comment (mustiikhalil) [inline todoie/Bindable.swift]: This is a normal value, with a did set function. so when the value is changed the observer is invoked.

File: todoie/Bindable.swift [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+        didSet {
+            observer?(value)
+        }
+    }
+
+    /**
+    binds the code from the ViewController to the ViewModel
+     - Parameters:
+        - observer: clousre the will be called when the value is changed
+     */
```
Comment (mustiikhalil) [inline todoie/Bindable.swift]: This function just takes the code of closure and sets it to the observer

File: todoie/Controllers/LoginViewController.swift [hunk 5]
```diff
@@ -22,34 +27,87 @@ class LoginViewController: UIViewController {
...
+    }
+}
+
+//MARK:- Setup LoginViewModel
+
+extension LoginViewController {
+    
+    func setupLoginViewModelObservers() {
+        
+        // Binds the clousre with the object isUserLoggedIn
```
Comment (mustiikhalil) [inline todoie/Controllers/LoginViewController.swift]: viewModel has a bindable object of type isUserLoggedIn, and we bind the dismiss code to it

File: todoie/ViewModel/LoginViewModel.swift [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+//  Created by Mustafa Khalil on 1/2/19.
+//  Copyright © 2019 Aurora. All rights reserved.
+//
+
+import Foundation
+import Firebase
+import GoogleSignIn
+
+class LoginViewModel: NSObject {
+
```
Comment (mustiikhalil) [inline todoie/ViewModel/LoginViewModel.swift]: Bindable object with Type Bool

File: todoie/ViewModel/LoginViewModel.swift [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+        }
+        
+        // Network manager should generate the google user Credentials and saves them
+        NetworkManager.shared.generateGoogleUserCredentials(user: user) { (error) in
+            if let err = error {
+                print(err.localizedDescription)
+                self.isUserLoggedIn.value = false
+                return
+            }
+            // sets the isUserLoggedIn to true to notify the VC that this is done
```
Comment (mustiikhalil) [inline todoie/ViewModel/LoginViewModel.swift]: Changing the value here will invoke the observer because of the `didSet`

**Comment by suhabaobaid** [comment]
@mustiikhalil the code I understood, just what are you proposing to use this observer for? which is the logic of binding the view model with the controller. is it going to be a generic observer?

**Pull request merged**